pkg/controller/controller_utils_test.go

```diff
@@ -1359,3 +1584,158 @@ func TestAddOrUpdateTaintOnNode(t *testing.T) {
 			test.name, test.requestCount, test.nodeHandler.RequestCount)
 	}
 }
+
+// BenchmarkPodNamespaceLabelIndexer tests the performance of single label lookup vs full list scan
+func BenchmarkPodNamespaceLabelIndexer(b *testing.B) {
+	const numPods = 150000
+
+	// Create a shared index informer
+	informer := cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return &v1.PodList{}, nil
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return watch.NewFake(), nil
+			},
+		},
+		&v1.Pod{},
+		0,
+		cache.Indexers{},
+	)
+
+	if err := AddPodNamespaceLabelIndexer(informer); err != nil {
+		b.Fatalf("failed to add indexer: %v", err)
+	}
+
+	labelValues := []string{"a", "b", "c", "d", "e", "f", "g", "h", "i", "j"}
+
+	type labelGenMode int
+	const (
+		uniformMode labelGenMode = iota
+		zipfMode
+	)
+
+	runBenchmark := func(mode labelGenMode) {
+		b.Helper()
+		informer = cache.NewSharedIndexInformer(
+			&cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+					return &v1.PodList{}, nil
+				},
+				WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+					return watch.NewFake(), nil
+				},
+			},
+			&v1.Pod{},
+			0,
+			cache.Indexers{},
+		)
+		if err := AddPodNamespaceLabelIndexer(informer); err != nil {
+			b.Fatalf("failed to add indexer: %v", err)
+		}
+
+		// Setup Zipf generator
+		r := rand.New(rand.NewSource(42))
+		var zipf *rand.Zipf
+		if mode == zipfMode {
+			zipf = rand.NewZipf(r, 1.2, 1, uint64(len(labelValues)-1))
+		}
+		largeString := strings.Repeat("x", 10000)
+		// Create pods with a single label "app"
+		for i := 0; i < numPods; i++ {
+			var v string
+			if mode == zipfMode {
+				v = labelValues[zipf.Uint64()]
+			} else {
+				v = labelValues[r.Intn(len(labelValues))]
+			}
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace:   "default",
+					Name:        fmt.Sprintf("pod-%d", i),
+					UID:         types.UID(fmt.Sprintf("uid-%d", i)),
+					Labels:      map[string]string{"app": v},
```

Comment:
This benchmark should the perf diff when there's 1 label.
It'd be good to also include 5 labels and 10 labels.